---
title: "Mycobacterium gene module clustering"
output: html_notebook
---

******
#Reading and Resources

1. Yen Yi's MRes report [Large Scale Computational Analysis of Coding and Non-coding Element Expression in Mycobacterium tuberculosis](Mycobacteria_project/HBXV7_AMS_MSCI_DISSERTATION_SUBMISSION.pdf)

2. Rosanna's MRes report [Computational analysis of combined RNA-seq datasets to predict novel ncRNAs in Mycobacterium tuberculosis](Mycobacteria_project/MRes_dissertation_RJ.pdf)


## work on gathering/annotating list of ncRNAs in mtb

***literature search***

Papers that identify ncRNA in mtb

1. [Cortes T, Schubert OT, Rose G, Arnvig KB, Comas I, Aebersold R, Young DB. 2013. Genome-wide mapping of transcriptional start sites defines an extensive leaderless transcriptome in Mycobacterium tuberculosis. Cell Rep. 5:1121–1131. ](http://dx.doi.org/10.1016/ j.celrep.2013.10.031)

-TSS validation

2. [Arnvig KB, Comas I, Thomson NR, Houghton J, Boshoff HI, Croucher NJ, Rose G, Perkins TT, Parkhill J, Dougan G, Young DB. 2011. Sequence-based analysis uncovers an abundance of non-coding RNA in the total transcriptome of Mycobacterium tuberculosis. PLoS Pathog. 7:e1002342. ](http://dx.doi.org/10.1371/journal.ppat.1002342)

3. [Miotto P, Forti F, Ambrosi A, Pellin D, Veiga DF, Balazsi G, Gennaro ML, Di Serio C, Ghisotti D, Cirillo DM. 2012. Genome-wide discovery of small RNAs in Mycobacterium tuberculosis. PLoS One 7:e51950. ](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0051950) 

4. [Dejesus, M. A. et al. Comprehensive essentiality analysis of the Mycobacterium tuberculosis genome via saturating transposon mutagenesis. MBio 8, 1–17 (2017).](https://mbio.asm.org/content/8/1/e02133-16)

  -esp S5(UTRs and promoters), S4 (predicted sRNAs)
  
5. [Shell SS, Wang J, Lapierre P, Mir M, Chase MR, Pyle MM, Gawande R, Ahmad R, Sarracino DA, Ioerger TR, Fortune SM, Derbyshire KM, Wade JT, Gray TA. 2015. Leaderless transcripts and small proteins are common features of the mycobacterial translational landscape. PLoS Genet11:e1005641. doi:10.1371/journal.pgen.1005641](https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1005641)

-TSS validation

6. [Arnvig KB, Young DB. 2009. Identification of small RNAs in Mycobacte- rium tuberculosis. Mol Microbiol 73:397–408. ](https://onlinelibrary.wiley.com/doi/full/10.1111/j.1365-2958.2009.06777.x)

7. [DiChiara JM, Contreras-Martinez LM, Livny J, Smith D, McDonough KA, Belfort M. 2010. Multiple small RNAs identified in Mycobacterium bovis BCG are also expressed in Mycobacterium tuberculosis and Mycobacterium smegmatis. ](https://academic.oup.com/nar/article/38/12/4067/2409235)

8. [Wang M, Fleming J, Li Z, Li C, Zhang H, Xue Y, Chen M, Zhang Z, Zhang XE, Bi L. 2016. An automated approach for global identification of sRNA-encoding regions in RNA-Seq data from Mycobacterium tubercu- losis. Acta Biochim Biophys Sin (Shanghai) 48:544–553. ](https://academic.oup.com/abbs/article/48/6/544/2195062)

9. [Mycobrowser website lists 62 putative ncRNA genes](https://mycobrowser.epfl.ch/home/about)

  can't download the list! use only for looking up particular ncRNAs

10. [Rfam database lists 22 sRNAs, 12 with structure](https://rfam.xfam.org/search?q=mycobacterium%20tuberculosis%20AND%20rna_type:%22sRNA%22%20AND%20TAXONOMY:%2283332%22)


11. [Gerrick, E. R. et al. Small RNA profiling in mycobacterium tuberculosis identifies mrsi as necessary for an anticipatory iron sparing response. Proc. Natl. Acad. Sci. U. S. A. 115, 6464–6469 (2018). ](https://www.pnas.org/content/115/25/6464)

  saved S1 list of putative predicted sRNAs using 'BSfinder' (same as Dejesus?)
  
12. [Dinan, A. M. et al. Relaxed Selection Drives a Noisy Noncoding Transcriptome in Members of the. 5, 1–9 (2014). ](https://mbio.asm.org/content/5/4/e01169-14)


Should I start with Gerrick et al and Tine's published list (Cortes), UTR's from Shell, et al and then use essentiality info from DeJesus? look at baerhunter predictions and Tine's unpublished list

Definitely: name (using 'new' system with ncRv1 as prefix), any old name that is used in literature, start, stop, strand, references, 'verified by' (for example RACE, northern blot, etc), 5' TSS


other info: expression conditions measured, flanking genes (though this is basically included in name and position above), RFAM, log2fold change in different conditions

start with first set. Should I have list of references and legend, or use PMID? might be easier for our use if have references listed.

Naming protocol using:

[Lamichhane, G., Arnvig, K. B. & McDonough, K. A. Definition and annotation of (myco)bacterial non-coding RNA. Tuberculosis 93, 26–29 (2013).](https://doi.org/10.1016/j.tube.2012.11.010)

Will focus on Arnvig list of ncRNAs. Cross-check Gerrick's (and DeJesus's?) computationally predicted ncRNAs with validated TSS sites from Shell and Cortes papers.

use union of TSSs in both conditions in Cortes paper

need to find TSS that lies in 10 nt before/within gene boundary and include actual TSS coordinate in df for Gerrick annotations

file is:  Gerrick_ncRNA_TSS_positions.txt


perhaps create an overlap index (Jaccard index) to figure out how much the genomic ranges overlap between sets 

[The BSGatlas: An enhanced annotation of genes and transcripts for the Bacillus subtilis genome with improved information access ](https://www.biorxiv.org/content/10.1101/807263v1.full)


also, Irilenia gave me file of highly expressed ncRNAs from Rosanna's mRes: "/Mycobacteria_project/combined_h37rv_Mtb_high_expressed.gff3""

but decided to run baerhunter again using Yen-yi's .bam files:
/d/projects/u/zchayyt/MTB_sample_accession_files/sample_accessions/

accessions_M.txt   PRJNA278760_22_Safa  QC_original         test_bam
Local_scripts      PRJNA327080_15       QC_postBWAmem
PRJEB19976_15p_SE  PRJNA390669_12       QC_postfastp
PRJEB65014_3       PRJNA507615_12_SE    QC_posttrimmomatic


install baerhunter on thoth R library

```{bash}
module load R/v4.0.1
```

```{R}
install.packages("devtools")
#Warning in install.packages("devtools") :
#  'lib = "/s/software/R/v4.0.1/lib64/R/library"' is not writable
#Would you like to use a personal library instead? (yes/No/cancel) yes
#Would you like to create a personal library
#‘~/R/x86_64-pc-linux-gnu-library/4.0’
#to install packages into? (yes/No/cancel) No
#Error in install.packages("devtools") : unable to install packages


install.packages("devtools")
#Warning in install.packages("devtools") :
#  'lib = "/s/software/R/v4.0.1/lib64/R/library"' is not writable
#Would you like to use a personal library instead? (yes/No/cancel) yes
#Would you like to create a personal library
#‘~/R/x86_64-pc-linux-gnu-library/4.0’
#to install packages into? (yes/No/cancel) yes

```

should I rename this library?

also installed baerhunter and all dependencies (see baerhunter_trial.R)

```{R}
library(baerhunter)
library(GenomicAlignments)
library(IRanges)
library(Rsamtools)
library(Rsubread)
library(DESeq2)

feature_file_editor(bam_directory="/d/projects/u/zchayyt/MTB_sample_accession_files/sample_accessions/PRJNA278760_22_Safa",
                    original_annotation_file="MtbH37RvNC_000962.3.gff",
                    annot_file_dir = "/d/in16/u/sj003/refseqs/Mtb/",
                    output_file="/d/projects/u/sj003/output_BH_12_10/PRJNA278760_22.gff3",
                    original_sRNA_annotation="ncRNA",
                    low_coverage_cutoff=5, #cutoff for noise
                    high_coverage_cutoff=10,
                    min_sRNA_length=40,
                    min_UTR_length=50,
                    paired_end_data=TRUE,
                    strandedness="stranded")

```

make R script calling above and call in bash:

```{bash}
nohup R CMD BATCH $my_path/scripts/BH_make_gff.R &
```

this is not really working. says 'nohup ignoring input'
tried again without nohup--taking forever

not working no matter what i do, same error:

[1] "Extracted plus strand data from BAM files"
Error in h(simpleError(msg, call)) : 
  error in evaluating the argument 'i' in selecting a method for function '[': 'match' requires vector arguments
Calls: feature_file_editor -> sRNA_calc -> [ -> %in% -> .handleSimpleError -> h
Execution halted


I'm thinking maybe it doesn't work with R version (4.0.1). I'm using R version 3.5.1 (2018-07-02) on my machine. Try loading with that version (3.5.2 on thoth) and try again.

```{bash}
module load R/v3.5.2
```

re-load devtools--makes new personal R library for 3.5

/d/user6/sj003/R/x86_64-pc-linux-gnu-library/3.5

```{R}
install.packages("devtools")
library(devtools)
library(IRanges)
library(Rsamtools)
library(DESeq2)
library(Rsubread)
devtools::install_github("irilenia/baerhunter")
library(baerhunter)
```

i don't think the libraries are necessary except for the session used--have to re-load them before using baerhunter, but packages are in personal R library, or general library that R has access to

```{bash}
nohup R CMD BATCH $my_path/scripts/BH_make_gff.R >& BH_make_gff.Rout &
```

got a warning from server that I was exceeding my allocated disk space limits. killed process.
```{bash}
kill%2  ##this is job number
```

looks like it is related to using personal R library in the home directory. need to move this to /d/in16/u/sj003. Deleted 3.5 and 4.0 libraries in home directory.

make files in working directory:
```{bash}
touch .Renviron
touch R_packages
```
.Renviron contains line:  R_LIBS = /d/in16/u/sj003/R_packages

when I reload R and install devtools, this time it works using correct folder.

Irilenia pointed out that "strandedness" refers to +/- stranded datasets. need to confirm this for each library. all of Yen-yi's datasets were reversely stranded.

going to re-do using R in correct directory, and using smallest dataset, PRJEB65014. this corresponds to E-MTAB-6011 (using number of samples, can't look at fastq files to find this identifier). Looked up on sra, this is correct bio-project number.

| bio-project | dataset |
|-----------|------------|
| PRJEB65014 | E-MTAB-6011 |
| PRJNA278760 | GSE670035 |
| PRJNA327080 | GSE83814 |
| PRJNA390669 | GSE100097 |


folder for .bam files:
/d/projects/u/zchayyt/MTB_sample_accession_files/sample_accessions/

```{R}
library(baerhunter)
library(GenomicAlignments)
library(IRanges)
library(Rsamtools)
library(DESeq2)
library(Rsubread)

# create a directory to hold the output files..
if (!(dir.exists("./output_BH_13_10/"))) { dir.create("./output_BH_13_10/")
}
feature_file_editor(bam_directory="/d/projects/u/zchayyt/MTB_sample_accession_files/sample_accessions/PRJEB65014_3/BWA_mem/",
                    original_annotation_file="MtbH37RvNC_000962.3.gff",
                    annot_file_dir = "/d/in16/u/sj003/refseqs/Mtb/",
                    output_file="output_BH_13_10/PRJE65014_3.gff3",
                    original_sRNA_annotation="ncRNA",
                    low_coverage_cutoff=5, #cutoff for noise
                    high_coverage_cutoff=10,
                    min_sRNA_length=40,
                    min_UTR_length=50,
                    paired_end_data=TRUE,
                    strandedness="reversely_stranded")

## run with other datasets:
#/d/projects/u/zchayyt/MTB_sample_accession_files/sample_accessions/PRJNA278760_22_Safa/
# Done 13/10/20
# /d/projects/u/zchayyt/MTB_sample_accession_files/sample_accessions/PRJNA327080_15/BWA_mem/
# Done 13/10/20
# /d/projects/u/zchayyt/MTB_sample_accession_files/sample_accessions/PRJNA390669_12/BWA_mem/

```

OK, that worked! have to use step to create new directory and only include name of directory in working directory, not entire path. Viewed gff3 on artemis. Got gff3 files for all four datasets.


