---
title: "UTR_gene_assignment"
output: html_notebook
author: Jennifer J. Stiens
---
date: 29 April, 2021

Write script to assign utrs to genes.

Name of putative non-coding utr in gff: "putative_UTR.m210813_210891"
m=minus strand, p=plus strand followed by genome coordinates
Doesn't specify 3' or 5'

change to 3UTR.RvXXXXx or 5UTR.RvXXXXx

What if is in-between two genes? Choose upstream gene: 3UTR? 
Could double-check with TSS? If TSS present= 5UTR, no TSS = 3UTR.
Manually checked UTRs from midnightblue: very few have adjacent genes segregated in the same module

```{r assign_utrs}
# use genomic ranges to assign gene name to UTRs
library(GenomicRanges)
library(rtracklayer)
library(stringr)
#BiocManager::install("GenomicFeatures")
#library(GenomicFeatures)
#define the GRanges object subject as the collection of genes 
# read in gff for mtb (use same one as for BH prediction) as genomic ranges object
mtb_granges<-import("~/git/mtb_modules/ref_seqs/Mtb_h37rv.ASM19595v2_AL123456.3.gff3")
# filter for type='gene'
mtb_granges<-mtb_granges[mcols(mtb_granges)$type=="gene"]
mtb_genes<-as.data.frame(mtb_granges)

Rv3249c<-mtb_granges[mcols(mtb_granges)$ID=="gene:Rv3249c"]
Rv3249c
# 3629752-3630387 is range on - strand
# look for 3' utr, putative utr following 3629752

#open file of saved genes from module
KME_filtered_darkred<-readLines("~/git/mtb_modules/module_genes/darkred_all_genes.txt")
gene_df <- KME_filtered_darkred
refseq_name <- "AL123456.3"
tss_file <- "~/git/mtb_modules/Data/shell_cortes_srna_tss.txt"
#create granges objects for all the utrs in list of module genes
# parse list of genes for start and end coordinates

  # returns a dataframe with each predicted utr given an assigment of 5' 3' or NA
  UTRs <- gene_df[grepl("UTR", gene_df)]
  utrs <- str_split_fixed(UTRs, "UTR.", 2)[,2]
  UTR_df           <- data.frame(matrix(0, nrow=length(utrs), ncol=4),stringsAsFactors = F)
  colnames(UTR_df) <- c("pred_utr", "start", "stop", "strand")
  UTR_df$pred_utr  <- UTRs
  # parse coordinates and strand from utr name
  for (i in 1:length(utrs)){
    strand <- substr(utrs[i], 1, 1)
    start  <- str_split_fixed(utrs[i], "_", 2)[,1]
    start  <- sub(".", "", start)
    stop   <- str_split_fixed(utrs[i], "_", 2)[,2]
    UTR_df$strand[i] <- strand
    UTR_df$start[i]  <- as.numeric(start)
    UTR_df$stop[i]   <- as.numeric(stop)
  }
  UTR_df$strand <- sub('p', '+', UTR_df$strand)
  UTR_df$strand <- sub('m', '-', UTR_df$strand)
  #View(UTR_df)
  # create grange object of utr coordinates
  utr_Granges<-GRanges(seqnames = refseq_name,
                         ranges   = IRanges(UTR_df$start, 
                                            end = UTR_df$stop),
                         strand   = Rle(strand(UTR_df$strand))
                         )
  # searching whether utr co-locates with gene?
  # precede is gene that "is preceded" by the UTR (comes after UTR==downstream)
  pg<-GenomicRanges::precede(utr_Granges, mtb_granges,ignore.strand=F)
  utr_precede_genes<-mtb_genes$gene_id[pg]
  # follow is gene that 'is followed' by UTR (comes before UTR==upstream)
  ug<-GenomicRanges::follow(utr_Granges, mtb_granges, ignore.strand=F)
  utr_follow_genes<-mtb_genes$gene_id[ug]
  # nearest(utrs, refseq, select="all", ignore.strand = F) will give me nearest gene feature.
  # select=all means returns both nearest features in case of a tie, arbitrary chooses one.
  # see how many ties
  a<-nearest(utr_Granges, mtb_granges, select="all", ignore.strand=F)
  b<-as.matrix(a)
  #show frequency of hits for each utr
  c<-as.data.frame(table(b[,1]))
  utr_tied<-c$Freq
  # see nearest gene (arbitrarily choose in ties)
  ng<-nearest(utr_Granges, mtb_granges, select="arbitrary", ignore.strand=F)
  utr_nearest_genes<-mtb_genes$gene_id[ng]
  #length(utr_nearest_genes)


# #add to utr df, preceding and following genes, and start coordinate
# utr_df_darkred<-data.frame(UTRs)
# utr_df_darkred$start  <-UTR_df$start
# utr_df_darkred$end    <-UTR_df$stop
# utr_df_darkred$strand <-UTR_df$strand
# have to consider strand for this? did precede() function consider strand? yes
UTR_df$downstream <-utr_precede_genes
UTR_df$upstream   <-utr_follow_genes
UTR_df$nearest    <-utr_nearest_genes
UTR_df$tied       <-utr_tied
#View(UTR_df)
# if nearest == precede, 3' UTR; if nearest ==follow, 5'UTR
# if tied? need to look for tss

# find TSS in utrs
# how many of srna-seq are verified with TSS at start??
# subset darkredGranges by strand (is this necessary?)
utr_fwd_Granges<-utr_Granges[strand(utr_Granges) == "+"]
fwd_df<-as.data.frame(utr_fwd_Granges)
utr_rev_Granges<-utr_Granges[strand(utr_Granges) =="-"]
rev_df<-as.data.frame(utr_rev_Granges)
# read in tss and make into granges obj
tss<-read.delim(tss_file, sep=" ")
tss_gr<-GRanges(seqnames = refseq_name,
                         ranges   = IRanges(tss$Genome.position, 
                                            end = tss$Genome.position),
                         strand   = Rle(strand(tss$Strand))
                         )
head(tss_gr)

## find overlap of each to tss within 10 bp of start 
# forward strand tss
fwd_tss_overlaps <-
  findOverlaps(
    utr_fwd_Granges,
    tss_gr,
    type = "start",
    maxgap = 10,
    select = "first",
    ignore.strand = FALSE
  )
qh<-as.matrix(fwd_tss_overlaps)
fwd_df$tss<-ifelse(is.na(qh), FALSE, tss[qh])
#View(fwd_df)

# reverse strand tss
rev_tss_overlaps<-
  findOverlaps(
    utr_rev_Granges,
    tss_gr,
    type = "end",
    maxgap = 10,
    select = "first",
    ignore.strand = FALSE
  )
qh<-as.matrix(rev_tss_overlaps)
rev_df$tss<-ifelse(is.na(qh), FALSE, tss[qh])

#put back together
df_tss<-rbind(fwd_df, rev_df)
# order by start
df_tss_o<-df_tss[order(df_tss$start),]
#nrow(df_tss_o)
#View(df_tss_o)

# add to dataframe
UTR_df$TSS <- df_tss_o$tss
#View(UTR_df)
# limit identified sRNAs and 5' UTRs to those with TSS (5' UTRs only) and 3' UTRs 
# that are upstream of a gene (finish 20-40 nt before downstream gene)

# I think I need to look at distance to nearest
#distanceToNearest: Returns the distance for each range in x to its nearest neighbor in the subject.
# or alternatively, use 'distance' and put in list of preceding and following genes?

#dist_to_darkred<-GenomicRanges::distanceToNearest(utr_Granges, mtb_granges, ignore.strand=F)
# all distances are 0, so must be adjacent to something (or between two genes)
# confirms that if tied, is between two genes and should be excluded, unless it has a TSS
# doesn't this exclude possible 3' UTRs that run between consecutive genes?

# put in start coordinate for each gene into dataframe from gff (end coordinate if on - strand)
# dataframe of gene_id, start (use end for neg strand)

# list of preceding gene_ids: utr_precede_genes, utr_follow_genes
#mtb_granges_df<-data.frame(mtb_granges)
mtb_genes$ID<-sub("gene:", "", mtb_genes$ID)
#View(mtb_genes)

# only need start of downstream gene to see if 3' UTR (end of UTR > 20nts from downstream start)
for (i in 1:nrow(UTR_df)){
  if (UTR_df$strand[i] == "+"){
    UTR_df$downstream_start[i]  <-
      mtb_genes$start[match(UTR_df$downstream[i], mtb_genes$ID)]
    UTR_df$dist_to_start[i] <-
      UTR_df$downstream_start[i] - UTR_df$stop[i]
  }else{
    # for minus strand, need end coordinate of downstream gene
    UTR_df$downstream_start[i]  <-
      mtb_genes$end[match(UTR_df$downstream[i], mtb_genes$ID)]
    UTR_df$dist_to_start[i] <-
      UTR_df$start[i] - UTR_df$downstream_start[i]
  }
}
# assign 5', 3' or NA to each utr
for (i in 1:nrow(UTR_df)){
  # if tied (same dist between two genes, located between two genes)
  if (UTR_df$tied[i] == 2){
    # check for TSS
    if (UTR_df$TSS[i] != FALSE){
        UTR_df$utr[i] <- "5'UTR"
    }else{
        UTR_df$utr[i] <- "NA"
    }
  #not tied (closer to either upstream or downstream gene)
  }else{
    if (UTR_df$nearest[i] == utr_df_darkred$upstream[i]){
      # check distance to downstream gene
        if (UTR_df$dist_to_start[i] > 40){
          UTR_df$utr[i] <- "3'UTR"
        }else{
          UTR_df$utr[i] <- "NA"
        }
    }else{
      if (UTR_df$nearest[i] == UTR_df$downstream[i]){
        # check if there is TSS 
        if (UTR_df$TSS[i] != FALSE){
          UTR_df$utr[i] <- "5'UTR"
        }else{
          UTR_df$utr[i] <- "NA"
        }
      }
    }
  }
}

View(UTR_df)
```


This has eliminated about 50% of utrs in module I looked at. I wonder if i shouldn't have done this before using all of the utrs in the wgcna--could have filtered this way previous to analysis. Maybe re-do? 

Could 5' UTRs exist without known TSS? TSS's change in different conditions--how are they detected?

This method is very conservative, probably some 3' UTRs between close genes (in operons) that are eliminated here, and also 5' UTRs without TSSs. So maybe ok they are used in network analysis.

Make function for UTR assignment. Also, maybe make list of UTRs from extended gff, and include sRNAs with TSS?


```{r test_function, include=FALSE}
source("~/git/mtb_modules/Scripts/utr_assign.R", local = knitr::knit_global())
# or sys.source("your-script.R", envir = knitr::knit_global())

# needed to save each function in a different script and source separately for it to work--something about packages?
source("~/git/mtb_modules/Scripts/tss_srnas.R", local = knitr::knit_global())

mod_color <- "brown"

mod_genes <- colnames(analysis[,moduleColors2==mod_color])

mod_utrs <- utr_assign(mod_genes, 
                    refseq_name = "AL123456.3",
                    annot_file="~/git/mtb_modules/ref_seqs/Mtb_h37rv.ASM19595v2_AL123456.3.gff3",
                    tss_file="~/git/mtb_modules/Data/shell_cortes_srna_tss.txt") 
# find srnas with tss within 20 nt
mod_srnas <- tss_srnas(mod_genes, 
                      refseq_name = "AL123456.3",
                      #annot_file="~/git/mtb_modules/ref_seqs/Mtb_h37rv.ASM19595v2_AL123456.3.gff3",
                      tss_file="~/git/mtb_modules/Data/shell_cortes_srna_tss.txt") 

# make new list with genes and new utr labels
# remove UTRs from list of genes and replace with categorised UTRs
rem_ncrnas  <- mod_genes[!grepl("UTR|sRNA", mod_genes)]
find_srnas  <- mod_srnas$TSS[which(mod_srnas$TSS != "FALSE")]
find_utrs   <- mod_utrs$utr[which(mod_utrs$utr != "NA")]
mod_genes2  <- c(rem_ncrnas, find_utrs, find_srnas)
mod_genes_file <- paste("~/git/mtb_modules/module_genes/", mod_color, "_all_genes.txt", sep="")
write(mod_genes2, mod_genes_file)

```

I should generate one of these files for each module. I can filter genes in module after? Curious to see if any of the putative srnas that I've excluded have large gene significance/module membership?
Maybe change so TSS verified srnas are indicated and others still remain?

```{r get_module_genes}

source("~/git/mtb_modules/Scripts/utr_assign.R", local = knitr::knit_global())
source("~/git/mtb_modules/Scripts/tss_srnas.R", local = knitr::knit_global())

for (i in 1:48){
  mod_color  <- colnames(MEs2)[i]
  color_name <- substr(mod_color, 3, nchar(mod_color))
  mod_genes  <- colnames(analysis[,moduleColors2==color_name])

  mod_utrs <- utr_assign(mod_genes, 
                    refseq_name = "AL123456.3",
                    annot_file="~/git/mtb_modules/ref_seqs/Mtb_h37rv.ASM19595v2_AL123456.3.gff3",
                    tss_file="~/git/mtb_modules/Data/shell_cortes_srna_tss.txt") 
  mod_srnas <- tss_srnas(mod_genes, 
                      refseq_name = "AL123456.3",
                      tss_file="~/git/mtb_modules/Data/shell_cortes_srna_tss.txt") 

  # make new list with genes and new utr labels
  # remove UTRs from list of genes and replace with categorised UTRs
  rem_ncrnas  <- mod_genes[grepl("gene", mod_genes)]
  find_utrs   <- mod_utrs$utr[which(mod_utrs$utr != "NA")]
  mod_genes2  <- c(rem_ncrnas, find_utrs, mod_srnas$pred_srna)
  mod_genes_file <- paste("~/git/mtb_modules/module_genes/", color_name, "_all_genes.txt", sep="")
  write(mod_genes2, mod_genes_file)
}

```

Looking at artemis, some of the sRNAs look like they could possibly be 5'UTRs? I've extended TSS requirement to 20nt because based on which bam I look at, peak sometimes slightly offset.

